### PR TITLE
Remove static_cast

### LINKED ISSUE #2307: Change the static_assert in logging to use `atomic<Logger::Level>` to reduce uses of `int`.
https://github.com/Azure/azure-sdk-for-cpp/blob/92839dcc5681961946d093555f9b8b9dfc96d2e5/sdk/core/azure-core/inc/azure/core/internal/diagnostics/log.hpp#L17

To avoid use of `int` here, consider changing this to `atomic<Logger::Level>`.

Recommended by @BillyONeal 
> Yes it means that before C++17 you can't assert atomic<Logger::Level>::is_always_lock_free but the runtime assert() is probably OK

cc @antkmsft

### PR DESCRIPTION
**Pull request by RickWinter**
Fixes #2307

### CODE REVIEW

**Review by antkmsft — CHANGES_REQUESTED** [review]
We should not be removing that. `atomic<int>` can be lock-free, and we can check for that. We can't check whether `atomic<Logger::Level>` is lock-free in C++14. You are removing a safety mechanism designed to guarantee use of lock-free ints.

**Comment by antkmsft** [comment]
In case you are trying to make it compile on a platform where the static_assert doesn't compile, removing the lock-free int check is not the right fix. If a platform does not have a lock-free int, we must either write a special version that relies on a smaller lock-free type, or on a lock, or a code that disables logging on such rare platform.

**Comment by RickWinter** [comment]
The static_cast is unnecessary.  I can add back a check that the underlying type is an int.

**Comment by antkmsft** [comment]
@RickWinter

g_logLevel HAS to be atomic<int>. Given that, comparison and initialization mixing of int and enum is not going to compile:
```
atomic<int> g_logLevel = ...;
Logger::Level level = ...;

level >= g_logLevel; // error: no match for ‘operator>’ (operand types are ‘enum Logger::Level’ and ‘atomic<int>’)
g_logLevel = Logger::Level::Warning; // error: cannot convert ‘enum Logger::Level’ to ‘atomic<int>’ in initialization
```

File: sdk/core/azure-core/inc/azure/core/internal/diagnostics/log.hpp [hunk 1]
```diff
@@ -11,24 +11,24 @@
...
+        std::is_same<int, std::underlying_type<Logger::Level>::type>::value == true,
         "Logger::Level values must be representable as lock-free");
 
+    static_assert(ATOMIC_INT_LOCK_FREE == 2, "atomic<int> must be lock-free");
+
     static_assert(ATOMIC_BOOL_LOCK_FREE == 2, "atomic<bool> must be lock-free");
 
     static AZ_CORE_DLLEXPORT std::atomic<bool> g_isLoggingEnabled;
-    static AZ_CORE_DLLEXPORT std::atomic<LogLevelInt> g_logLevel;
+    static AZ_CORE_DLLEXPORT std::atomic<Logger::Level> g_logLevel;
```
Comment (antkmsft) [inline sdk/core/azure-core/inc/azure/core/internal/diagnostics/log.hpp:23]: Still, no explicit guarantee that if `atomic<int>` is lock-free, that `atomic<Logger::Level>` is also lock-free, even though the underlying type is `int`.

**Comment by antkmsft** [comment]
Nowhere in the standard it says that `std::atomic<T>` where `sizeof(T) == sizeof(int)` HAS to use the same template specialization. It is pretty reasonable to assume, but it does not say so.
And why remove that. We explicitly use `int`, and that is why it is better implementation, and `static_cast` removal does not compare in value.

There's nothing that would prevent implementations from doing this:
```
#include <iostream>

enum class Enum : int { A, B, C };

template <typename T>
struct Otomic
{
    static const constexpr bool IsLockFree = false;
};

template <>
struct Otomic<int>
{
    static const constexpr bool IsLockFree = true;
};

int main()
{
    std::cout << "Otomic<Enum>::IsLockFree: " << Otomic<Enum>::IsLockFree << "\n";
    std::cout << "Otomic<int>::IsLockFree: " << Otomic<int>::IsLockFree << "\n";
}
```

Output:
```
Otomic<Enum>::IsLockFree: 0
Otomic<int>::IsLockFree: 1
```

File: sdk/core/azure-core/inc/azure/core/internal/diagnostics/log.hpp [hunk 2]
```diff
@@ -11,24 +11,30 @@
...
-    using LogLevelInt = std::underlying_type<Logger::Level>::type;
+    static_assert(
+        std::is_same<int, std::underlying_type<Logger::Level>::type>::value == true,
+        "Logger::Level values must be representable as lock-free");
 
+    // Ensure that the LogLevel is of size that is compatible with cmpxchg16b
+    // If this is true then a reasonable implementation of atomic will be lock_free
     static_assert(
-        std::is_same<int, LogLevelInt>::value == true && ATOMIC_INT_LOCK_FREE == 2,
+        sizeof(Logger::Level) <= 8 && (sizeof(Logger::Level) & (sizeof(Logger::Level) - 1)) == 0,
```
Comment (antkmsft) [inline sdk/core/azure-core/inc/azure/core/internal/diagnostics/log.hpp:21]: Remove this assertion - it is hard to read, it is an overkill, and it assumes way too much, about `cmpxchg16b`, but what about ARMs? They don't have this instruction. No, if we decide to assume that an enum with underlying type of int is stored exactly as atomic int, plus we check that int is atomic, then the other `static_assert`s here are sufficient, this one is not necessary. I'd say this assert is harder to read than a static_cast.

**Comment by antkmsft** [comment]
https://github.com/Azure/azure-sdk-for-cpp/issues/2307#issuecomment-916313877

This, to me, reverses the meaning of #2307 being necessary to do in the first place.

**Comment by antkmsft** [comment]
Billy:
>>Wait, I read this as the opposite - that before C++17, there is no way to be guaranteed that atomic<Logger::Level> is atomic, so we are limited to int, for which we can guarantee.
>
>That's right. is_always_lock_free was added in '17. It is not necessarily true, particularly on implementations that chose the "base classes" version of atomic, that atomic is in any way related to atomic<Logger::Level>.

https://github.com/Azure/azure-sdk-for-cpp/issues/2307#issuecomment-916388721

**Comment by RickWinter** [comment]
The extra typedef and static_casts are not necessary.  That code is defending against terrible QoI which is a waste of time.  If the underlying type is an int but the implementation chooses to use some other implementation than atomic<int> the implementation is likely to have much bigger problems.

**Comment by antkmsft** [comment]
@RickWinter, can you at least remove the `sizeof(Logger::Level) <= 8 && (sizeof(Logger::Level) & (sizeof(Logger::Level) - 1)) == 0` assertion? We do still check that the underlying type is int, and we expect that the implementation will pick the same implementation of atomic internally. Given all that, no need for this additional check sizeof being power of two.

**Review by antkmsft — APPROVED** [review]

**Pull request merged**